url47e6aca666_pocorall_automaton_tgz-pJ8-net_pocorall_automaton_DatatypesJ8/src/net/pocorall/automaton/RegExp.java

```diff
@@ -0,0 +1,877 @@
+		return pos < b.length();
+	}
+
+	private char next() throws IllegalArgumentException {
+		if (!more())
+			throw new IllegalArgumentException("unexpected end-of-string");
+		return b.charAt(pos++);
+	}
+
+	private boolean check(int flag) {
+		return (flags & flag) != 0;
+	}
+
+	final RegExp parseUnionExp() throws IllegalArgumentException {
+		RegExp e = parseInterExp();
+		if (match('|'))
+			e = makeUnion(e, parseUnionExp());
+		return e;
+	}
+
+	final RegExp parseInterExp() throws IllegalArgumentException {
+		RegExp e = parseConcatExp();
+		if (check(INTERSECTION) && match('&'))
+			e = makeIntersection(e, parseInterExp());
+		return e;
+	}
+
+	final RegExp parseConcatExp() throws IllegalArgumentException {
+		RegExp e = parseRepeatExp();
+		if (more() && !peek(")|") && (!check(INTERSECTION) || !peek("&")))
+			e = makeConcatenation(e, parseConcatExp());
+		return e;
+	}
+
+	final RegExp parseRepeatExp() throws IllegalArgumentException {
+		RegExp e = parseComplExp();
+		while (peek("?*+{")) {
+			if (match('?'))
+				e = makeOptional(e);
+			else if (match('*'))
+				e = makeRepeat(e);
+			else if (match('+'))
+				e = makeRepeat(e, 1);
+			else if (match('{')) {
+				int start = pos;
+				while (peek("0123456789"))
+					next();
+				if (start == pos)
+					throw new IllegalArgumentException("integer expected at position " + pos);
+				int n = Integer.parseInt(b.substring(start, pos));
+				int m = -1;
+				if (match(',')) {
+					start = pos;
+					while (peek("0123456789"))
+						next();
+					if (start != pos)
+						m = Integer.parseInt(b.substring(start, pos));
+				} else
+					m = n;
+				if (!match('}'))
+					throw new IllegalArgumentException("expected '}' at position " + pos);
+				if (m == -1)
+					e = makeRepeat(e, n);
+				else
+					e = makeRepeat(e, n, m);
+			}
+		}
+		return e;
+	}
+
+	final RegExp parseComplExp() throws IllegalArgumentException {
+		if (check(COMPLEMENT) && match('~'))
+			return makeComplement(parseComplExp());
+		else
+			return parseCharClassExp();
+	}
+
+	final RegExp parseCharClassExp() throws IllegalArgumentException {
+		if (match('[')) {
+			boolean negate = false;
+			if (match('^'))
+				negate = true;
+			RegExp e = parseCharClasses();
+			if (negate)
+				e = makeIntersection(makeAnyChar(), makeComplement(e));
+			if (!match(']'))
+				throw new IllegalArgumentException("expected ']' at position " + pos);
+			return e;
+		} else
+			return parseSimpleExp();
+	}
+
+	final RegExp parseCharClasses() throws IllegalArgumentException {
+		RegExp e = parseCharClass();
+		while (more() && !peek("]"))
+			e = makeUnion(e, parseCharClass());
+		return e;
+	}
+
+	final RegExp parseCharClass() throws IllegalArgumentException {
+		char c = parseCharExp();
+		if (match('-'))
+			if (peek("]"))
+				return makeUnion(makeChar(c), makeChar('-'));
+			else
+				return makeCharRange(c, parseCharExp());
+		else
+			return makeChar(c);
+	}
+
+	final RegExp parseSimpleExp() throws IllegalArgumentException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 28. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.